pipeline/runtime/step_executor.go

```diff
@@ -110,7 +113,122 @@ func (e *StepExecutor) PollStep(ctx context.Context, r *api.PollStepRequest) (*a
 	return convertStatus(status), nil
 }
 
+func (e *StepExecutor) StreamOutput(ctx context.Context, r *api.StreamOutputRequest) (oldOut []byte, newOut <-chan []byte, err error) {
+	id := r.ID
+	if id == "" {
+		err = &errors.BadRequestError{Msg: "ID needs to be set"}
+		return
+	}
+
+	var stepOutput *StepOutput
+
+	// the runner will call this function just before the call to start step, so we wait a while for the step to start
+	for ts := time.Now(); ; {
+		e.mu.Lock()
+		stepOutput = e.stepOutput[id]
+		e.mu.Unlock()
+
+		if stepOutput != nil {
+			break
+		}
+
+		const timeoutDelay = 5 * time.Second
```

Comment:
If a step takes a millisecond to complete, it will take atleast 5 seconds in drone since 5 seconds is the minimum time after which logs are sent back.

Reply:
You misunderstood the code. The 5s delay is to abort streaming (if the step has not started in that period), not to start streaming...
Well, I admit it's a hacky solution, but it's there just to ensure that step is started just prior to the streaming. The two requests are started simultaneously by the runner...